### ISSUE TITLE
LibJS: m_ptr[offset] >> 6 == 2 assertion failed in Utf8CodepointIterator in less than comparison

### ISSUE THREAD

**Issue by Lubrsi** [SerenityOS/serenity#3549]
Found whilst fuzzing
File: [crash.txt](https://github.com/SerenityOS/serenity/files/5249012/crash.txt)

Trace:
```
USERSPACE(24) ASSERTION FAILED: m_ptr[offset] >> 6 == 2
../AK/Utf8View.cpp:210
[js(24:24)]: Terminating js(24) due to signal 6
[FinalizerTask(3:3)]: CPU[0] NP(error) fault at invalid address V0x01801d9c
[FinalizerTask(3:3)]: copy_from_user(0xc094a9ac, 0x01801d9c, 4) failed at V0x01801d9c
[FinalizerTask(3:3)]: 0xdeadc0de  Kernel::Processor::switch_context(Kernel::Thread*&, Kernel::Thread*&) +181
0xdeadc0de  Kernel::Scheduler::context_switch(Kernel::Thread*) +183
0xdeadc0de  Kernel::Scheduler::pick_next() +1593
0xdeadc0de  Kernel::Processor::check_invoke_scheduler() +125
0xdeadc0de  Kernel::Thread::die_if_needed() +157
0xdeadc0de  syscall_handler +1414
0xdeadc0de  syscall_asm_entry +49
0x080baa78  raise +22
0x080480bd  abort +13
0x080b6acb  tolower +0
0x080d5468  AK::Utf8CodepointIterator::operator*() const +180
0x0809384a  JS::abstract_relation(JS::Interpreter&, bool, JS::Value, JS::Value) +591
0x08093dd9  JS::less_than(JS::Interpreter&, JS::Value, JS::Value) +49
0x0805a0ce  JS::BinaryExpression::execute(JS::Interpreter&, JS::GlobalObject&) const +644
0x08059e81  JS::BinaryExpression::execute(JS::Interpreter&, JS::GlobalObject&) const +55
0x08059c3a  JS::ExpressionStatement::execute(JS::Interpreter&, JS::GlobalObject&) const +44
0x08066716  JS::Interpreter::execute_statement(JS::GlobalObject&, JS::Statement const&, AK::Vector<JS::Argument, 8ul>, JS::ScopeType) +414
0x0805c120  JS::ScopeNode::execute(JS::Interpreter&, JS::GlobalObject&) const +64
0x080654af  JS::Interpreter::run(JS::GlobalObject&, JS::Program const&) +313
0x0804b10b  parse_and_run(JS::Interpreter&, AK::StringView const&) +502
0x08048cba  main +2124
0x080498fe  _start +94
```

**Comment by bugaevc** [comment]
It seems `JS::Value::abstract_relation()` tries to decide strings as UTF-8, without ever checking that they are, in fact, valid UTF-8.

Now, I don't think JavaScript strings are *supposed* to be UTF-8. But even if they are, this is a suboptimal way to compare things. UTF-8 has this nice property that you can compare strings byte-by-byte (i.e. `strcmp()`) and it'll always yield the same result, so there's no need to do any decoding for that.

**Comment by linusg** [comment]
> Now, I don't think JavaScript strings are supposed to be UTF-8.

they should be UTF-16 [according to the spec](https://tc39.es/ecma262/#sec-terms-and-definitions-string-value), but are UTF-8 in LibJS.

And here's the relevant part of [Abstract Relational Comparison](https://tc39.es/ecma262/#sec-abstract-relational-comparison):

> 3. If `Type(px)` is String and `Type(py)` is String, then
>     a. If `IsStringPrefix(py, px)` is **true**, return **false**.
>     b. If `IsStringPrefix(px, py)` is **true**, return **true**.
>     c. Let `k` be the smallest nonnegative integer such that the code unit at index `k` within `px` is different from the code unit at index `k` within `py`. (There must be such a `k`, for neither String is a prefix of the other.)
>     d. Let `m` be the integer that is the numeric value of the code unit at index `k` within `px`.
>     e. Let `n` be the integer that is the numeric value of the code unit at index `k` within `py`.
>     f. If `m < n`, return **true**. Otherwise, return **false**.

And the current implementation in LibJS:

https://github.com/SerenityOS/serenity/blob/394e4c04cd5dfdca56f67e6384288a209bdbf6dc/Libraries/LibJS/Runtime/Value.cpp#L916-L939

**Comment by linusg** [comment]
This was fixed in #5901 - running the linked `crash.txt` code now correctly results in `false`.
